bson-kotlinx/src/main/kotlin/org/bson/codecs/kotlinx/BsonDecoder.kt

```diff
@@ -125,7 +126,6 @@ internal open class DefaultBsonDecoder(
                 return BsonArrayDecoder(reader, serializersModule, configuration)
             }
             is PolymorphicKind -> {
-                reader.readStartDocument()
```

Comment:
Can you explain the apparent lack of symmetry of the various branches in this method?  It's not clear to me why in some branches we have to call `readStartArray` or `readStartDocument`, but in others we do not.

Reply:
I've cleaned these up - now all the Decoders are responsible for using the reader and validating its state.